File: python/atelier_bench/metrics/overhead.py
"""overhead metric."""

from __future__ import annotations

from typing import TYPE_CHECKING

from ..contracts.metric_contract import MetricUnit
from ..support.abstract_metric import AbstractMetric

if TYPE_CHECKING:
    from ..dataset import Dataset


class OverheadMetric(AbstractMetric):
    KEY = "overhead"

    def __init__(self) -> None:
        super().__init__(self.KEY, MetricUnit.CENTS)

    def compute(self, data: "Dataset") -> float:
        return data.parts_cost_cents() * 0.15

File: go/atelier/metrics/overhead.go
// Package metrics carries the 16 metrics. Each embeds support.BaseMetric,
// declares only Compute, and registers itself in an init function.
package metrics

import (
	"atelier.example/lane/atelier/contracts"
	"atelier.example/lane/atelier/dataset"
	"atelier.example/lane/atelier/support"
)

// Compile-time proof of satisfaction; nothing else names the contract here.
var _ contracts.MetricContract = (*OverheadMetric)(nil)

// OverheadMetric computes the overhead metric.
type OverheadMetric struct {
	support.BaseMetric
}

// NewOverheadMetric builds the metric with its key and unit.
func NewOverheadMetric() *OverheadMetric {
	return &OverheadMetric{BaseMetric: support.NewBaseMetric("overhead", contracts.UnitCents)}
}

// Compute is the only contract method this type declares; Key and Unit are
// promoted from the embedded base.
func (m *OverheadMetric) Compute(data *dataset.Dataset) float64 {
	return float64(data.PartsCostCents()) * 0.15
}

// init registers this metric.
func init() {
	Register(NewOverheadMetric())
}
